File: cracking_the_coding_interview/10-4.cpp
#include <iostream>
#include <vector>
using namespace std;

class MyArray{
public:
    MyArray(vector<unsigned int> input){
        v = input;
    }
    int elementAt(int i){
        if(i < v.size()){
            return v[i];
        }else{
            return -1;
        }
    }
private:
    vector<unsigned int> v;
};

int findIndex(MyArray& myArray, int x, int left, int right){
    if(left >= right){
        return -1;
    }
    int mid = (left + right) / 2;
    if(myArray.elementAt(mid) == x){
        return mid;
    }else if(myArray.elementAt(mid) == -1 || myArray.elementAt(mid) > x){
        return findIndex(myArray, x, left, mid);
    }
    return findIndex(myArray, x, mid+1, right);
}

// constにしたらelementAt呼べなくなったのはなぜ
int findIndex(MyArray& myArray, int x){
    int i = 1;
    while(myArray.elementAt(i) != -1){
        i *= 2;
    }
    return findIndex(myArray, x, 0, i);
}

int main(void){
    vector<unsigned int> v = {0, 1, 2, 8, 10, 100, 726};
    MyArray myArray = MyArray(v);
    cout << findIndex(myArray, 10) << endl; // 4
    cout << findIndex(myArray, 3) << endl; // -1
    return 0;
}

File: cracking_the_coding_interview/10-4.py
import unittest

class Array:
    def __init__(self, array=[]):
        self.array = sorted(array)
    
    def element_at(self, i):
        if not self.array or len(self.array) <= i:
            return -1
        return self.array[i]
    
    def size(self):
        return len(self.array)

def find_index(nums, x):
    if nums.element_at(0) == -1:
        return None

    if nums.element_at(0) == x:
        return 0
    
    l, r = find_range(nums, x)
    return binary_search(nums, l, r, x)

def find_range(nums, target):
    i = 1
    l = i
    while nums.element_at(i) >= 0 and nums.element_at(i) < target:
        l = i
        i *= 2
    return l, binary_search_range(nums, l, i, target)

def binary_search_range(nums, left, right, target):
    if left > right:
        return right
    mid = (left + right) // 2
    if nums.element_at(mid) >= 0 and nums.element_at(mid + 1) == -1:
        return mid + 1

    if nums.element_at(mid) > target:
        return mid
    
    if nums.element_at(mid) == -1:
        return binary_search_range(nums, left, mid, target)
    else:
        return binary_search_range(nums, mid + 1, right, target)

def binary_search(nums, left, right, target):
    if left > right:
        return None
    
    mid = (left + right) // 2
    if nums.element_at(mid) == target:
        return mid
    elif nums.element_at(mid) > target:
        return binary_search(nums, left, mid, target)
    else:
        return binary_search(nums, mid + 1, right, target)

class Test(unittest.TestCase):
    def test_simple(self):
        nums = [0, 1, 2, 3, 4, 5]
        array = Array(nums)
        for i in range(array.size()):
            self.assertEqual(find_index(array, i), i)
        
        self.assertEqual(find_index(array, array.size()), None)

if __name__ == "__main__":
    unittest.main()